### PR TITLE
SM-12 Logging

### PR DESCRIPTION
**Pull request by valerio8787**

### CODE REVIEW

File: src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+            new IntrospectionProcessor(),
+        ];
+    }
+
+    /**
+     * @return \Monolog\Handler\StreamHandler
+     */
+    protected function createStreamHandler(): StreamHandler
+    {
+        $streamHandler = new StreamHandler(
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php:70]: Do we have preference for all these instantiations? Should't we keep this configurable from the outside? Meaning that we should inject this configuration.

File: src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+            $this->getLogFilePath(),
+            $this->loggerSettings->getVerboseLevel()
+        );
+        $formatter = new LogstashFormatter(static::NAME);
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php:74]: Will this logger remain hardwired to this format?

File: src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        return [
+            new PsrLogMessageProcessor(),
+            new IntrospectionProcessor(),
+        ];
+    }
+
+    /**
+     * @return \Monolog\Handler\StreamHandler
+     */
+    protected function createStreamHandler(): StreamHandler
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php:68]: The name indicates this should be in a module factory.

File: src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        $formatter = new LogstashFormatter(static::NAME);
+        $streamHandler->setFormatter($formatter);
+
+        return $streamHandler;
+    }
+
+    /**
+     * @return \Monolog\Handler\StreamHandler
+     */
+    protected function createConsoleStreamHandler(): StreamHandler
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php:83]: The name indicates this should be in a module factory.

File: src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php [hunk 5]
```diff
@@ -13,26 +14,28 @@ class ProcessFacade extends AbstractFacade implements ProcessFacadeInterface
     /**
      * @param array $payload
      * @param \SprykerMiddleware\Zed\Process\Business\Mapper\Map\MapInterface $map
+     * @param \Psr\Log\LoggerInterface $logger
      *
      * @return array
      */
-    public function map(array $payload, MapInterface $map)
+    public function map(array $payload, MapInterface $map, LoggerInterface $logger): array
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php:21]: I have to hand around the logger everywhere myself? 🤔

File: src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php [hunk 6]
```diff
@@ -19,18 +24,24 @@
  */
 class ProcessCommunicationFactory extends AbstractCommunicationFactory
 {
+    use LoggerTrait;
```
Comment (lazystar) [inline src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php:27]: What is this used for?

Comment (valerio8787) [reply src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php:27]: @lazystar LoggerTrait is used for getting of  configured Logger. getLogger() method came from it

File: src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php (hunk 5 above)
Comment (valerio8787) [reply src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php:21]: @lazystar it is horizontal dependency, we discussed it with @vlunov-spryker, and decided that we haven't another choice

File: src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php (hunk 3 above)
Comment (valerio8787) [reply src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php:68]: @lazystar It is a basic logger configuration that provided only as example. Each process can provide own logger configuration. In this case i think that possibility to configure all logger parameters in one place can do it less complicated, than injecting a lot of dependencies (processors, formatter, handlers and Config).

File: src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+use Monolog\Handler\StreamHandler;
+use Monolog\Processor\IntrospectionProcessor;
+use Monolog\Processor\PsrLogMessageProcessor;
+use Spryker\Shared\Config\Config;
+use Spryker\Shared\Log\Config\LoggerConfigInterface;
+use Spryker\Shared\Log\LogConstants;
+
+class MiddlewareLoggerConfig implements LoggerConfigInterface
+{
+    const NAME = 'SprykerMiddleware';
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php:16]: CHANNEL_NAME?

File: src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php (hunk 3 above)
Comment (PhilinTv) [reply src/SprykerMiddleware/Zed/Process/Business/Log/Config/MiddlewareLoggerConfig.php:68]: @valerio8787 then the name should be MonologLoggerConfig.
This implementation is too specific. So then it should be clarified as a spec-impl or become more generic (through factory, adapters etc).

File: src/SprykerMiddleware/Zed/Process/Business/Mapper/Mapper.php [hunk 8]
```diff
@@ -54,19 +65,38 @@ public function map(array $payload): array
...
+    protected function mapCallable(array $result, array $payload, string $key, callable $value): array
+    {
+        $mappedValue = $value($payload, $key);
+        $this->logger->debug('Mapping', [
+            'operation' => 'Map callable',
+            'new_key' => $key,
+            'old_key' => $value,
+            'data' => $mappedValue,
+        ]);
+        return $this->payloadManager->setValue($result, $key, $mappedValue);
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Mapper/Mapper.php:97]: Please don't use setter prefix (set-) to modify passing values. Setter should return void or $this. The reason is that PayloadManager is implemented in pseudo-OOP way: as a container for static functions.

File: src/SprykerMiddleware/Zed/Process/Business/Pipeline/Stage/Stage.php [hunk 9]
```diff
@@ -11,20 +12,39 @@ class Stage implements StageInterface
...
 
     /**
      * @inheritdoc
      */
     public function __invoke($payload)
     {
-        return $this->stagePlugin
-            ->process($payload);
+        $this->logger->info('Input Data', [
+            'stage' => get_class($this->stagePlugin),
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Pipeline/Stage/Stage.php:36]: Could you wrap it in a protected function?

File: src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php (hunk 5 above)
Comment (PhilinTv) [reply src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php:21]: Facade methods should take/return only primitives or transfer objects. Dependencies through facade are now allowed. 
@valerio8787  could you use factory?

File: src/SprykerMiddleware/Zed/Process/Business/ProcessFacadeInterface.php [hunk 10]
```diff
@@ -2,25 +2,28 @@
...
 
+use Psr\Log\LoggerInterface;
 use SprykerMiddleware\Zed\Process\Business\Mapper\Map\MapInterface;
 
 interface ProcessFacadeInterface
 {
     /**
      * @param array $payload
      * @param \SprykerMiddleware\Zed\Process\Business\Mapper\Map\MapInterface $map
+     * @param \Psr\Log\LoggerInterface $logger
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/ProcessFacadeInterface.php:13]: Please mark facade methods as API.

File: src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php [hunk 11]
```diff
@@ -13,26 +14,28 @@ class ProcessFacade extends AbstractFacade implements ProcessFacadeInterface
...
         return $this->getFactory()
-            ->createMapper($map)
+            ->createMapper($map, $logger)
             ->map($payload);
     }
 
     /**
      * @param array $payload
      * @param array $dictionary
+     * @param \Psr\Log\LoggerInterface $logger
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/ProcessFacade.php:31]: Please inherit doc-block from the interface.

File: src/SprykerMiddleware/Zed/Process/Business/Translator/Translator.php [hunk 12]
```diff
@@ -113,6 +121,30 @@ protected function translateByRule(array $result, array $payload, string $key, $
...
+     * @param callable $translation
+     *
+     * @return array
+     */
+    protected function translateCallable(array $result, array $payload, string $key, callable $translation): array
+    {
+        $inputValue = $this->payloadManager->getValueByKey($payload, $key);
+        $resultValue = $translation($inputValue, $key, $payload);
+        $this->logger->debug(
+            'Translation',
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Business/Translator/Translator.php:137]: Const?

File: src/SprykerMiddleware/Zed/Process/Communication/Console/ProcessConsole.php [hunk 13]
```diff
@@ -52,6 +55,13 @@ protected function configure()
             InputOption::VALUE_OPTIONAL,
             'Count of items that should be processed'
         );
+
+        $this->addOption(
+            static::OPTION_LOG_LEVEL,
+            'f',
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Communication/Console/ProcessConsole.php:61]: Const. Eg OPTION_LOG_LEVEL_SHORTCUT = 'f'

File: src/SprykerMiddleware/Zed/Process/Communication/Console/ProcessConsole.php [hunk 14]
```diff
@@ -104,4 +111,52 @@ protected function hasError()
...
+    /**
+     * @param \Symfony\Component\Console\Input\InputInterface $input
+     * @param \Generated\Shared\Transfer\ProcessSettingsTransfer $processSettingsTransfer
+     *
+     * @return void
+     */
+    protected function setIteratorOptions(InputInterface $input, ProcessSettingsTransfer $processSettingsTransfer): void
+    {
+        $processSettingsTransfer->setIteratorSettings(new IteratorSettingsTransfer());
+        $offset = $input->getOption(self::OPTION_ITERATOR_OFFSET) ?: 0;
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Communication/Console/ProcessConsole.php:124]: static?

File: src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php (hunk 6 above)
Comment (PhilinTv) [reply src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php:27]: @valerio8787 please make sure there is no business logic in the factory. And move logger BL to a delegated business model.

File: src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php [hunk 15]
```diff
@@ -153,4 +167,36 @@ protected function getRegisteredPostProcessorsList(): array
...
+        if (isset($configuration[$processSettingsTransfer->getName()])) {
+            return $configuration[$processSettingsTransfer->getName()]($processSettingsTransfer->getLoggerSettings());
+        }
+        return $this->createMiddlewareLogConfig($processSettingsTransfer->getLoggerSettings());
+    }
+
+    /**
+     * @return array
+     */
+    protected function getProcessLoggerConfigList(): array
```
Comment (PhilinTv) [inline src/SprykerMiddleware/Zed/Process/Communication/ProcessCommunicationFactory.php:188]: Configuration should come from a special delegated class - ProcessConfig => AbstractBundleConfig

**Review by PhilinTv — CHANGES_REQUESTED** [review]
Please fix/answer comments.

**Pull request merged**